Create a manifest of graphic elements in output

ashley-hebler opened the pull request:
### Changed
- `templates/graphic/project.config.js`, `templates/feature/project.config.js` - Adds new key for adding customization to the parsing step. For now the only one there is one to ignore Apple News for specified files.
- `templates/__common__/_package.json` - Adds puppeteer and a new `npm run parse` step
- `templates/feature/README.md`, `templates/graphic/README.md` - Adds new context for these changes
- `templates/graphic/app/templates/base.html` - New meta tags on graphics to be used internally.

### Added
- ` templates/__common__/_.npmrc`- This tells puppeteer to skip downloading chrome each time we create a new project. Saves us some storage on our machines.
- `templates/__common__/config/scripts/parse.js ` - Task runner that spins up a local server and kicks off the graphics-meta.js step.
- `templates/__common__/config/tasks/graphics-meta.js `- The base for this whole process. At a high level, this will step through the whole project /dist folder and extract all the relevant metadata info we specify. This is also where we build the JSON for Apple News and capture screenshots of graphics.


## Test steps:

1. Check out `graphic-metadata` and run `your/path/to/data-visuals-create/bin/data-visuals-create graphic my-test-graphic`
2. In that new project, add some sample HTML to the index.html. Feel free to use something from an old project or just this [test HTML](https://gist.github.com/ashley-hebler/44632a2c7eef5064b4d2b602dd100b97) I've been using.
3. Run `npm run serve` to preview the graphic. If you don't use the sample code, be sure to add a .button CTA to demonstrate how those will be handled in a screenshot.
3. Run `npm run build` then `npm run parse`

-- 🆕 steps since https://github.com/texastribune/data-visuals-create/pull/68/commits/3d8031b13408788977632dc8a05efc11c0c1af23 --
5. Confirm a **`manifest.json`** file is created. This is a list of the metatdata for each graphic.
6. Confirm a small and large preview .png is also created

## Changes to dist output after running `npm run parse`
**before:**
![Screen Shot 2021-01-27 at 3 11 36 PM](https://user-images.githubusercontent.com/2974713/106054909-47347f00-60b2-11eb-856a-4aa0c5fd8820.png)
**after:**
![Screen Shot 2021-01-27 at 3 12 29 PM](https://user-images.githubusercontent.com/2974713/106054907-456abb80-60b2-11eb-8bb4-404f690f61a8.png)




## Notes:
@mandicai - ~I didn't add the tags to the project config yet, but we can add as many new keys there as we see fit.~ 
2/4/21 Update: tags are now there. I just cherry-picked this commit https://github.com/texastribune/data-visuals-create/pull/34/commits/3a53493957c7818d4ec0dc2dec1d1e328ea32998

 I know there's quite a bit to review so take your time. I hope it's not too messy! Let me know if you want to talk through anything.

Also for now, I left this off of the build step, but it might make sense to make it more automatic and make [this line](https://github.com/texastribune/data-visuals-create/blob/a26700093e6afc0012d24d10f10b3e5c84104f1f/templates/__common__/_package.json#L7):
`"build": "NODE_ENV=production node ./config/scripts/build.js && npm run parse",`
once we verify it the process is working for us.

ashley-hebler commented:
test comment

ashley-hebler commented:
test again

mandicai commented:
test x 3

mandicai commented:
hello again

mandicai commented:
more testing

mandicai commented:
test

mandicai commented:
all the tests

ashley-hebler commented on templates/feature/README.md:
```diff
@@ -112,6 +116,14 @@ Projects created with `data-visuals-create` support two of the built-in ways tha
 
 The interface with Google Drive within `data-visuals-create` projects currently only supports using Oauth2 credentials to speak to the Google APIs. This requires a set of OAuth2 credentials that will be used to generate and save an access token to your computer. `data-visuals-create` projects have hardcoded locations for the credential file and token file, but you may override those with environm ...
 
+### Chrome
```
Sounds good to me! It'd technically work for features if they contain the `.graphic` class somewhere, but you're right. I think this just adds extra noise since we'll probably mostly use them in actual graphics only projects. Removed in https://github.com/texastribune/data-visuals-create/pull/68/commits/63c07debf3cfc3250d1ac13046633e72ec8417a0 ✅

ashley-hebler commented on templates/feature/project.config.js:
```diff
@@ -93,4 +93,22 @@ module.exports = {
+   * ],
+   *
+   */
+  parserOptions: {
```
@mandicai, that's correct! Since `app/bananas.html` ultimately compiles to `/dist/bananas/index.html`, it'll skip that bananas folder per this [line](https://github.com/texastribune/data-visuals-create/blob/5261bbe2e18d509bb77f9564904571aadb57be46/templates/__common__/config/tasks/graphics-meta.js#L245). Alternatively, adding `bananas/index.html` to that project.config array would do the same.

ashley-hebler commented on templates/feature/README.md:
```diff
@@ -100,6 +100,10 @@ The `workspace` directory is for storing all of your analysis, production and ra
 
 Pulls any `workspace` files that have been pushed to S3 back down to the project's local `workspace` directory. This is helpful for ensuring you're in sync with another developer.
 
+### `npm run parse`
```
@mandicai yes, great catch! I added a fix for this in https://github.com/texastribune/data-visuals-create/pull/68/commits/4314e89b0166d9b16cffbf891089773d1e1caa6c

ashley-hebler commented on templates/feature/project.config.js:
```diff
@@ -93,4 +93,22 @@ module.exports = {
    *
    */
   customFilters: {},
+  /**
```
Done! https://github.com/texastribune/data-visuals-create/pull/68/commits/a2818d8b440cf1fb00100c692b37dd810a867c7e

ashley-hebler commented on templates/__common__/config/tasks/graphics-meta.js:
```diff
@@ -0,0 +1,362 @@
+  const outputPath = path.join(paths.appDist, parentDir, '/');
+
+  // begin puppeteer navigation
+  const page = await browser.newPage();
```
Puppeteer seriously opens so many doors. It's a pretty heavy dependency, but I love it! Not installing chrome (hence the `CHROME_INSTALL_PATH` var) makes it a little less bulky at least.

ashley-hebler commented on templates/__common__/config/tasks/graphics-meta.js:
```diff
@@ -0,0 +1,362 @@
+};
+
+// path where chrome is typically installed on MacOS
+const CHROME_INSTALL_PATH =
```
perhaps! In my mind, it'd be pretty consistent for us since we're all on macs, but that is probably very short-sighted/OS discriminating of me 😆. Do you think it's a blocker? I do see [solves](https://github.com/hughsk/chrome-location/blob/master/index.js) out there in the world we could investigate

mandicai replied:
Hahah I saw similar npm packages! Could we just steal their raw code lololol, I was hesitant to add another package. I think I'm just wary about new Mac updates, and whether that could change paths on the machine, but maybe that's not very likely?

mandicai replied:
Can confirm that `chrome-location` does find the correct location: `/Applications/Google Chrome.app/Contents/MacOS/Google Chrome`

mandicai replied:
... Does anyone at the Tribune use a Windows or did we just shame all of them into using Mac's

ashley-hebler replied:
I'm in favor of making it more OS aware in a new PR since this one has grown to be so giant 😅

ashley-hebler commented:
@mandicai, believe it or not, I'm actually going to close this 😆. I chatted with Ty today about the graphics CMS plugin and I want to make some changes based on that.

ashley-hebler commented:
Ok @mandicai, we're back on track. I ended up cutting the step to generate the Apple News Format JSON. I think we can do that in the CMS based on what we'd already have available with this. I added different test steps based on the new output. Hopefully, there are fewer moving parts now 😅

mandicai commented:
Awesome, I'll go through and test those steps again! @ashley-hebler

mandicai commented:
Confirming that I can see the `manifest.json` and two images for my `index.html` graphic, but I'm not seeing them for my `static` graphic — I don't see those files inside the `dist/static/` folder.

ashley-hebler commented:
@mandicai sounds good on the first part. For some reason, I thought we used the static folder for images and non html files, but I think I was confusing it with the assets folder. I reverted the commit where I added a line to ignore that folder (https://github.com/texastribune/data-visuals-create/pull/68/commits/543dff546e68a8643846100f59863f0219977742) and that should work now.

mandicai commented:
Slack convo 1/28/2021 with @ashley-hebler 
- [ ] Chrome install path — some hesitation here, in case the path is different on different versions of Mac or if a team member uses Windows. Mandi will get some DV team members to test this PR. If a helpful error message is printed that prompts the user to fix their path, it might be enough.
- [ ] Does `npm run parse` work for apps pages? If we add it the `build` command, it will get applied to both features and graphics. Ashley confirmed that it does work for apps pages, the `manifest-json` file will be an empty array if there are no `graphic-` CSS classes in the HTML. Mandi will add some comments above the `graphic-` CSS classes in the HTML to let people know they shouldn't remove this.
- [ ] Add `{% set graphicTitle = 'foo' %}` and `{% set graphicDesc = 'Description of graphic for screenshot accessibility' %}` to our graphic templates. For Illustrator graphics, we typically set the title in the Illustrator file itself and not the HTML, so we'd need to add it to the HTML for it to appear in `manifest.json`.

ashley-hebler commented:
@mandicai, I'm responding to your checklist above

>  Chrome install path...

Sounds good! I think if everything with that path goes smoothly, I'd love to shelve the more sophisticated path retrieval logic for a later PR – just to keep this one at the size it is now.

 
> Mandi will add some comments above the `graphic-` CSS classes in the HTML to let people know they shouldn't remove this...

So I attempted to tackle this here:
- [data-graphic](https://github.com/texastribune/data-visuals-create/blob/graphic-metadata/templates/graphic/app/index.html#L18)
- [data-title](https://github.com/texastribune/data-visuals-create/blob/graphic-metadata/templates/graphic/app/index.html#L20)
- [data-source & data-credit](https://github.com/texastribune/data-visuals-create/blob/graphic-metadata/templates/graphic/app/index.html#L27)

 > Add `{% set graphicTitle = 'foo' %}` and `{% set graphicDesc = 'Description of graphic for screenshot accessibility' %}` to our graphic templates...
 
 And then similar to above, in the static file, I added this [note](https://github.com/texastribune/data-visuals-create/blob/graphic-metadata/templates/graphic/app/static.html#L15). I also used your language for that part in the [graphics-meta readme](https://github.com/texastribune/data-visuals-create/blob/graphic-metadata/templates/graphic/graphics-meta.md#full-example-with-ai2html-graphic)
 
 Also feel free to cut/edit anything from that readme or move it. It's probably a little long-winded.

I'm going to leave this alone for a bit. Let me know how team testing goes and as always, thank you for your patience on this!

mandicai commented:
@ashley-hebler This graphics metadata README is beautiful 😭

csessig86 commented:
LGTM

CarlaAstudillo commented:
This LGTM too! Thank you so much, this is fantastic!

darlacameron commented:
This is mostly good, but I found one bug. 

When I'm running `npm run serve`, my http://localhost:3001/ port is used by Browserify. That means that `npm run parse` fails with `Error: listen EADDRINUSE: address already in use :::3001`. If I stop the serve, it works fine.

ashley-hebler commented:
hey @darlacameron, good catch. I think of `npm run parse` as part of the final build step. We talked about even making part of `npm run build` eventually. You'd run it order to package the project up for the CMS. It's not really intended to be run in tandem with `npm run serve`. Does that make sense? @mandicai, curious about your thoughts on this.

ashley-hebler commented:
@darlacameron, I know yall are on deadline right now so no rush, but see if 4b0cf69 fixes that issue you found.

ashley-hebler commented on CHANGELOG.md:
```diff
@@ -10,6 +10,21 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+### Added
+- ` templates/__common__/_.npmrc`- This tells puppeteer to skip downloading chrome each time we create a new project. Saves us some storage on our machines.
+- `templates/__common__/config/scripts/parse.js ` - Task runner that spins up a local server and kicks off the graphics-meta.js step.
+- `templates/__common__/config/tasks/graphics-meta.js `- The base for this whole process. At a high level, this will step through the whole project /dist folder and extract all the relevant metadata info we specify. This is also where we build the JSON for Apple News and capture screenshots of graphics.
```
> This is also where we build the JSON for Apple News and capture screenshots of graphics.

So we're not actually generating the Apple News JSON anymore. On the CMS side, we're just going to use manifest.json to build the apple news JSON ourselves if that makes sense. I think this sentence can just be:

> This is also where we capture screenshots of graphics.
